pkg/dockerregistry/server/manifestservice.go

```diff
@@ -0,0 +1,274 @@
+		return "", regapi.ErrorCodeManifestInvalid.WithDetail(err)
+	}
+
+	_, err = m.manifests.Put(WithRepository(ctx, m.repo), manifest, options...)
+	if err != nil {
+		return "", err
+	}
+
+	// Calculate digest
+	dgst := digest.FromBytes(canonical)
+
+	// Upload to openshift
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: m.repo.namespace,
+			Name:      m.repo.name,
+		},
+		Image: imageapi.Image{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: dgst.String(),
+				Annotations: map[string]string{
+					imageapi.ManagedByOpenShiftAnnotation: "true",
+				},
+			},
+			DockerImageReference:         fmt.Sprintf("%s/%s/%s@%s", m.repo.registryAddr, m.repo.namespace, m.repo.name, dgst.String()),
+			DockerImageManifest:          string(payload),
+			DockerImageManifestMediaType: mediaType,
+		},
+	}
+
+	for _, option := range options {
+		if opt, ok := option.(distribution.WithTagOption); ok {
+			ism.Tag = opt.Tag
+			break
+		}
+	}
+
+	if err = mh.FillImageMetadata(ctx, &ism.Image); err != nil {
+		return "", err
+	}
+
+	// Remove the raw manifest as it's very big and this leads to a large memory consumption in etcd.
+	ism.Image.DockerImageManifest = ""
+	ism.Image.DockerImageConfig = ""
+
+	if err = m.repo.registryOSClient.ImageStreamMappings(m.repo.namespace).Create(&ism); err != nil {
+		// if the error was that the image stream wasn't found, try to auto provision it
+		statusErr, ok := err.(*kerrors.StatusError)
+		if !ok {
+			context.GetLogger(ctx).Errorf("error creating ImageStreamMapping: %s", err)
+			return "", err
+		}
+
+		if quotautil.IsErrorQuotaExceeded(statusErr) {
+			context.GetLogger(ctx).Errorf("denied creating ImageStreamMapping: %v", statusErr)
+			return "", distribution.ErrAccessDenied
+		}
+
+		status := statusErr.ErrStatus
+		if status.Code != http.StatusNotFound ||
+			(strings.ToLower(status.Details.Kind) != "imagestream" /*pre-1.2*/ && strings.ToLower(status.Details.Kind) != "imagestreams") ||
+			status.Details.Name != m.repo.name {
+			context.GetLogger(ctx).Errorf("error creating ImageStreamMapping: %s", err)
+			return "", err
+		}
+
+		stream := imageapi.ImageStream{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: m.repo.name,
+			},
+		}
+
+		uclient, ok := UserClientFrom(m.ctx)
+		if !ok {
+			context.GetLogger(ctx).Errorf("error creating user client to auto provision image stream: Origin user client unavailable")
+			return "", statusErr
+		}
+
+		if _, err := uclient.ImageStreams(m.repo.namespace).Create(&stream); err != nil {
+			if quotautil.IsErrorQuotaExceeded(err) {
+				context.GetLogger(ctx).Errorf("denied creating ImageStream: %v", err)
+				return "", distribution.ErrAccessDenied
+			}
+			context.GetLogger(ctx).Errorf("error auto provisioning ImageStream: %s", err)
+			return "", statusErr
+		}
+
+		// try to create the ISM again
+		if err := m.repo.registryOSClient.ImageStreamMappings(m.repo.namespace).Create(&ism); err != nil {
```

Comment:
We should either squash this call with the one above that creates the image stream (add the extra info in the stream prior to creation) or at the very least retry on 404 here.